tests/test_audit.py

```diff
@@ -200,3 +210,181 @@ def test_audit_end_to_end(session, users, groups, http_client, base_url, graph):
+        for groupname in ("group-1", "group-2", "group-3", "group-4", AUDITORS_GROUP)
+    }
+    # create users
+    users.update({
+        username + '@a.co': User.get_or_create(session, username=username + '@a.co')[0]
+        for username in ("user11", "user12", "user13", "user14",
+                         "user21", "user22", "user23",
+                         "user41", "user42", "user43",
+        )
+    })
+    # create permissions
+    permissions.update({
+        permission: Permission.get_or_create(
+            session, name=permission, description="{} permission".format(permission)
+        )[0]
+        for permission in [PERMISSION_NAME]
+    })
+    # add users to groups
+    for (groupname, username, role) in (("group-1", "user11", "owner"),
+                                        ("group-1", "user12", "member"),
+                                        ("group-1", "user13", "np-owner"),
+                                        ("group-1", "user14", "owner"),
+                                        ("group-2", "user13", "np-owner"),
+                                        ("group-2", "user21", "owner"),
+                                        ("group-2", "user22", "member"),
+                                        ("group-3", "user12", "owner"),
+                                        ("group-3", "user22", "owner"),
+                                        ("group-4", "user21", "np-owner"),
+                                        ("group-4", "user41", "member"),
+                                        ("group-4", "user42", "owner"),
+                                        ("group-4", "user43", "np-owner"),
+                                        ):
+        add_member(groups[groupname], users[username + "@a.co"], role=role)
+    # add group-4 as member of group-1
+    add_member(groups["group-1"], groups["group-4"])
+    # add user14 to auditors group
+    add_member(groups[AUDITORS_GROUP], users["user14@a.co"])
+    # grant permissions to groups
+    #
+    # give the test permission to groups 1 and 2, and group 4 should
+    # also inherit from group 1
+    grant_permission(groups["group-1"], permissions[PERMISSION_NAME])
+    grant_permission(groups["group-2"], permissions[PERMISSION_NAME])
+    grant_permission(groups[AUDITORS_GROUP], permissions[PERMISSION_AUDITOR])
+
+    session.commit()
+    graph.update_from_db(session)
+    # done setting up
+
+    # now a few pre-op checks
+    assert not graph.get_group_details('group-1').get(AUDITED_GROUP)
+    assert not graph.get_group_details('group-4').get(AUDITED_GROUP)
+    assert get_users(graph, AUDITORS_GROUP) == set(["user14@a.co"])
+    assert get_users(graph, "group-3") == set(["user12@a.co", "user22@a.co"])
+
+    #
+    # run the promotion logic -> nothing should happen because the
+    # test-permission is not yet audited
+    #
+    background = BackgroundProcessor(settings, None)
+    background.promote_nonauditors(session)
+    session.commit()
```

Comment:
Should be redundant (called at the end of promote_nonauditors) I believe? Also for the other call

Reply:
done